File: R/linkedin_analytics.R
#!/usr/bin/env Rscript

#' LinkedIn Analytics Data Analysis
#' 
#' This script analyzes LinkedIn analytics data exported from your profile.
#' 
#' Requirements:
#' - tidyverse
#' - readxl
#' - lubridate
#' - ggplot2
#' - scales
#' 
#' Usage:
#' 1. Export your LinkedIn analytics data as an XLSX file
#' 2. Update the file_path variable below to point to your XLSX file
#' 3. Run the script: Rscript linkedin_analytics.R

# Load required libraries
if (!require("tidyverse")) install.packages("tidyverse")
if (!require("readxl")) install.packages("readxl")
if (!require("lubridate")) install.packages("lubridate")
if (!require("ggplot2")) install.packages("ggplot2")
if (!require("scales")) install.packages("scales")

library(tidyverse)
library(readxl)
library(lubridate)
library(ggplot2)
library(scales)

# Set file path to your LinkedIn analytics XLSX file
file_path <- "LinkedIn_Analytics.xlsx"

# Function to load all sheets from Excel file
load_analytics_data <- function(file_path) {
  tryCatch({
    # Get all sheet names
    sheet_names <- excel_sheets(file_path)
    
    # Create a named list to store each sheet as a dataframe
    data <- list()
    
    for (sheet in sheet_names) {
      data[[sheet]] <- read_excel(file_path, sheet = sheet)
      cat(sprintf("Loaded sheet: %s with %d rows\n", sheet, nrow(data[[sheet]])))
    }
    
    return(data)
  }, error = function(e) {
    cat(sprintf("Error loading data: %s\n", e$message))
    return(NULL)
  })
}

# Function to analyze profile views
analyze_profile_views <- function(profile_views_df) {
  if (is.null(profile_views_df) || nrow(profile_views_df) == 0) {
    return("No profile views data available for analysis.")
  }
  
  # Identify date column
  date_col <- names(profile_views_df)[grep("date", tolower(names(profile_views_df)))]
  
  if (length(date_col) > 0) {
    # Convert date column to Date type
    profile_views_df[[date_col[1]]] <- as.Date(profile_views_df[[date_col[1]]])
    
    # Group by date and count views
    views_by_date <- profile_views_df %>%
      group_by(!!sym(date_col[1])) %>%
      summarise(views = n())
    
    # Create a plot
    p <- ggplot(views_by_date, aes(x = !!sym(date_col[1]), y = views)) +
      geom_line(color = "#0077B5", size = 1) +
      geom_point(color = "#0077B5", size = 3) +
      labs(
        title = "LinkedIn Profile Views Over Time",
        x = "Date",
        y = "Number of Views"
      ) +
      theme_minimal() +
      theme(
        plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),
        axis.title = element_text(size = 12),
        axis.text = element_text(size = 10)
      )
    
    # Save the plot
    ggsave("profile_views_trend.png", p, width = 10, height = 6, dpi = 300)
    cat("Profile views trend chart saved as 'profile_views_trend.png'\n")
    
    # Basic statistics
    total_views <- nrow(profile_views_df)
    
    # Analyze viewer companies if available
    if ("Company" %in% names(profile_views_df)) {
      top_companies <- profile_views_df %>%
        count(Company, sort = TRUE) %>%
        head(10)
      
      cat("\nTop Companies Viewing Your Profile:\n")
      print(top_companies)
      
      # Create a bar chart of top companies
      p_companies <- ggplot(top_companies, aes(x = reorder(Company, n), y = n)) +
        geom_bar(stat = "identity", fill = "#0077B5") +
        coord_flip() +
        labs(
          title = "Top Companies Viewing Your Profile",
          x = "Company",
          y = "Number of Views"
        ) +
        theme_minimal()
      
      ggsave("top_companies.png", p_companies, width = 10, height = 6, dpi = 300)
      cat("Top companies chart saved as 'top_companies.png'\n")
    }
  }
}

# Function to analyze post engagement
analyze_post_engagement <- function(posts_df) {
  if (is.null(posts_df) || nrow(posts_df) == 0) {
    return("No post engagement data available for analysis.")
  }
  
  # Identify engagement columns (likes, comments, shares, etc.)
  engagement_patterns <- c("like", "comment", "share", "click", "impression")
  engagement_cols <- names(posts_df)[sapply(names(posts_df), function(col) {
    any(sapply(engagement_patterns, function(pattern) grepl(pattern, tolower(col))))
  })]
  
  if (length(engagement_cols) > 0) {
    # Calculate total engagement by post
    posts_df$total_engagement <- rowSums(posts_df[engagement_cols], na.rm = TRUE)
    
    # Sort by engagement
    top_posts <- posts_df %>%
      arrange(desc(total_engagement)) %>%
      head(5)
    
    cat("\nYour Top 5 Performing Posts:\n")
    for (i in 1:nrow(top_posts)) {
      post <- top_posts[i, ]
      content_col <- intersect(c("Content", "Post", "Text"), names(post))
      date_col <- intersect(c("Date", "Posted on"), names(post))
      
      cat(sprintf("%d. Date: %s\n", i, if(length(date_col) > 0) post[[date_col[1]]] else "N/A"))
      if (length(content_col) > 0) {
        content <- post[[content_col[1]]]
        if (nchar(content) > 100) content <- paste0(substr(content, 1, 100), "...")
        cat(sprintf("   Content: %s\n", content))
      }
      cat(sprintf("   Total Engagement: %d\n", post$total_engagement))
      
      engagement_breakdown <- sapply(engagement_cols, function(col) sprintf("%s: %d", col, post[[col]]))
      cat(sprintf("   Breakdown: %s\n\n", paste(engagement_breakdown, collapse = ", ")))
    }
    
    # Create visualization of engagement distribution
    engagement_summary <- posts_df %>%
      summarise(across(all_of(engagement_cols), sum, na.rm = TRUE)) %>%
      pivot_longer(cols = everything(), names_to = "Metric", values_to = "Count")
    
    p_engagement <- ggplot(engagement_summary, aes(x = reorder(Metric, Count), y = Count)) +
      geom_bar(stat = "identity", fill = "#0077B5") +
      coord_flip() +
      labs(
        title = "LinkedIn Engagement Metrics Distribution",
        x = "Engagement Type",
        y = "Count"
      ) +
      theme_minimal()
    
    ggsave("engagement_distribution.png", p_engagement, width = 10, height = 6, dpi = 300)
    cat("Engagement distribution chart saved as 'engagement_distribution.png'\n")
  }
}

# Main function
main <- function() {
  cat("LinkedIn Analytics Data Analyzer (R)\n")
  cat("==================================\n")
  
  # Load data
  data <- load_analytics_data(file_path)
  if (is.null(data)) {
    cat("Could not load analytics data. Please check the file path.\n")
    return()
  }
  
  # Analyze different aspects of the data based on sheet names
  for (sheet_name in names(data)) {
    cat(sprintf("\n\nAnalyzing %s...\n", sheet_name))
    df <- data[[sheet_name]]
    
    if (any(grepl("profile|view", tolower(sheet_name)))) {
      analyze_profile_views(df)
    } else if (any(grepl("post|article|engagement", tolower(sheet_name)))) {
      analyze_post_engagement(df)
    }
  }
  
  cat("\n\nThis analysis is powered by a basic implementation.\n")
  cat("For more advanced analytics, visit Professional Profile Analytics (PPA).\n")
}

# Run the main function
main()

File: Python/linkedin_analytics.py
#!/usr/bin/env python3
"""
LinkedIn Analytics Data Analysis
-------------------------------
This script analyzes LinkedIn analytics data exported from your profile.

Requirements:
- pandas
- matplotlib
- seaborn
- openpyxl

Usage:
1. Export your LinkedIn analytics data as an XLSX file
2. Update the file_path variable below to point to your XLSX file
3. Run the script: python linkedin_analytics.py
"""

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
import os

def load_analytics_data(file_path):
    """Load LinkedIn analytics data from XLSX file"""
    try:
        # Read all sheets from the Excel file
        xl = pd.ExcelFile(file_path)
        
        # Create a dictionary to store each sheet as a dataframe
        data = {}
        for sheet_name in xl.sheet_names:
            data[sheet_name] = pd.read_excel(file_path, sheet_name=sheet_name)
            print(f"Loaded sheet: {sheet_name} with {len(data[sheet_name])} rows")
        
        return data
    except FileNotFoundError:
        print(f"Error: File {file_path} not found.")
        return None
    except Exception as e:
        print(f"Error loading data: {e}")
        return None

def analyze_profile_views(profile_views_df):
    """Analyze profile views data"""
    if profile_views_df is None or len(profile_views_df) == 0:
        return "No profile views data available for analysis."
    
    # Convert date column to datetime if needed
    date_col = [col for col in profile_views_df.columns if 'date' in col.lower()][0]
    profile_views_df[date_col] = pd.to_datetime(profile_views_df[date_col])
    
    # Group by date and count views
    views_by_date = profile_views_df.groupby(date_col).size().reset_index(name='views')
    
    # Create a plot
    plt.figure(figsize=(12, 6))
    plt.plot(views_by_date[date_col], views_by_date['views'], marker='o')
    plt.title('LinkedIn Profile Views Over Time')
    plt.xlabel('Date')
    plt.ylabel('Number of Views')
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    
    # Save the plot
    plt.savefig('profile_views_trend.png')
    print("Profile views trend chart saved as 'profile_views_trend.png'")
    
    # Basic statistics
    total_views = profile_views_df.shape[0]
    unique_viewers = profile_views_df['Profile viewer'].nunique() if 'Profile viewer' in profile_views_df.columns else "N/A"
    
    # Analyze viewer companies if available
    if 'Company' in profile_views_df.columns:
        top_companies = profile_views_df['Company'].value_counts().head(10)
        print("\nTop Companies Viewing Your Profile:")
        print(top_companies)
    
    # Analyze viewer job titles if available
    if 'Job title' in profile_views_df.columns:
        top_titles = profile_views_df['Job title'].value_counts().head(10)
        print("\nTop Job Titles Viewing Your Profile:")
        print(top_titles)
    
    return {
        "total_views": total_views,
        "unique_viewers": unique_viewers
    }

def analyze_post_engagement(posts_df):
    """Analyze post engagement data"""
    if posts_df is None or len(posts_df) == 0:
        return "No post engagement data available for analysis."
    
    # Identify engagement columns (likes, comments, shares, etc.)
    engagement_cols = [col for col in posts_df.columns if any(metric in col.lower() 
                                                            for metric in ['like', 'comment', 'share', 'click', 'impression'])]
    
    # Calculate total engagement by post
    if engagement_cols:
        posts_df['total_engagement'] = posts_df[engagement_cols].sum(axis=1)
        
        # Sort by engagement
        top_posts = posts_df.sort_values('total_engagement', ascending=False).head(5)
        
        print("\nYour Top 5 Performing Posts:")
        for i, (idx, post) in enumerate(top_posts.iterrows(), 1):
            print(f"{i}. Date: {post.get('Date', 'N/A')}")
            print(f"   Content: {post.get('Content', post.get('Post', 'N/A'))[:100]}...")
            print(f"   Total Engagement: {post['total_engagement']}")
            print(f"   Breakdown: {', '.join([f'{col}: {post[col]}' for col in engagement_cols])}")
            print()
    
    # Create visualization of post performance over time
    if 'Date' in posts_df.columns and engagement_cols:
        posts_df['Date'] = pd.to_datetime(posts_df['Date'])
        plt.figure(figsize=(14, 7))
        
        for col in engagement_cols:
            plt.plot(posts_df['Date'], posts_df[col], marker='o', label=col)
            
        plt.title('LinkedIn Post Engagement Over Time')
        plt.xlabel('Date')
        plt.ylabel('Engagement Count')
        plt.legend()
        plt.grid(True, alpha=0.3)
        plt.tight_layout()
        
        # Save the plot
        plt.savefig('post_engagement_trend.png')
        print("Post engagement trend chart saved as 'post_engagement_trend.png'")

def main():
    print("LinkedIn Analytics Data Analyzer")
    print("===============================")
    
    # Update this path to your LinkedIn analytics XLSX file
    file_path = "LinkedIn_Analytics.xlsx"
    
    # Load data
    data = load_analytics_data(file_path)
    if not data:
        print("Could not load analytics data. Please check the file path.")
        return
    
    # Analyze different aspects of the data based on sheet names
    for sheet_name, df in data.items():
        print(f"\n\nAnalyzing {sheet_name}...")
        
        if any(term in sheet_name.lower() for term in ['profile', 'view']):
            results = analyze_profile_views(df)
            if isinstance(results, dict):
                print(f"Total profile views: {results['total_views']}")
                print(f"Unique viewers: {results['unique_viewers']}")
        
        elif any(term in sheet_name.lower() for term in ['post', 'article', 'engagement']):
            analyze_post_engagement(df)
    
    print("\n\nThis analysis is powered by a basic implementation.")
    print("For more advanced analytics, visit Professional Profile Analytics (PPA).")

if __name__ == "__main__":
    main()
